File: backend/app/schemas/ghost_virality.py
"""Pydantic schemas for the Ghost Virality pipeline."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


# ---------------------------------------------------------------------------
# Scout Jobs
# ---------------------------------------------------------------------------


class GhostScoutJobCreate(BaseModel):
    niche: str = Field(..., min_length=1, max_length=120, description="Content niche (e.g. 'fitness', 'finance')")
    competitor_accounts: list[str] = Field(default_factory=list, description="Instagram usernames to scout")


class GhostScoutJobOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    niche: str
    competitor_accounts: Optional[list] = None
    status: str
    celery_task_id: Optional[str] = None
    reels_scraped: int
    ghost_viral_found: int
    error_message: Optional[str] = None
    created_by: Optional[uuid.UUID] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    created_at: datetime


# ---------------------------------------------------------------------------
# Pattern Card
# ---------------------------------------------------------------------------


class PatternCardOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    ghost_post_id: uuid.UUID
    hook_duration_seconds: Optional[float] = None
    hook_clip_path: Optional[str] = None
    scene_cut_count: Optional[int] = None
    transcript_snippet: Optional[str] = None
    transcript_language: Optional[str] = None
    transcript_confidence: Optional[float] = None
    visual_text: Optional[str] = None
    audio_type: Optional[str] = None
    audio_id: Optional[str] = None
    audio_name: Optional[str] = None
    hook_archetype: Optional[str] = None
    raw_card: Optional[dict] = None
    created_at: datetime


# ---------------------------------------------------------------------------
# Ghost Viral Posts
# ---------------------------------------------------------------------------


class GhostViralPostOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    reel_id: str
    account_username: str
    niche: Optional[str] = None

    view_count: Optional[int] = None
    like_count: Optional[int] = None
    comment_count: Optional[int] = None
    follower_count: Optional[int] = None

    outlier_reach_ratio: Optional[float] = None
    ghost_virality_delta: Optional[float] = None
    like_percentile: Optional[float] = None
    view_velocity: Optional[float] = None
    like_velocity: Optional[float] = None

    strategy_label: Optional[str] = None
    comment_sentiment: Optional[str] = None

    permalink: Optional[str] = None
    thumbnail_url: Optional[str] = None
    posted_at: Optional[datetime] = None
    pattern_card_ready: bool
    detected_at: datetime

    pattern_card: Optional[PatternCardOut] = None


# ---------------------------------------------------------------------------
# Niche Overview
# ---------------------------------------------------------------------------


class NicheOverviewOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    niche: str
    sample_size: int
    p10: Optional[float] = None
    p30: Optional[float] = None
    p50: Optional[float] = None
    p70: Optional[float] = None
    p90: Optional[float] = None
    outlier_reach_threshold: float
    ghost_viral_count: Optional[int] = None
    updated_at: datetime


# ---------------------------------------------------------------------------
# Trial Reels
# ---------------------------------------------------------------------------


class TrialReelCreate(BaseModel):
    ghost_post_id: Optional[uuid.UUID] = None
    niche: Optional[str] = Field(None, max_length=120)
    variation_label: Optional[str] = Field(None, max_length=255)
    post_url: Optional[str] = Field(None, max_length=1024)
    notes: Optional[str] = None


class TrialReelUpdate(BaseModel):
    post_url: Optional[str] = Field(None, max_length=1024)
    views_at_1k: Optional[int] = None
    completion_rate: Optional[float] = Field(None, ge=0.0, le=1.0)
    likes: Optional[int] = None
    comments: Optional[int] = None
    shares: Optional[int] = None
    status: Optional[str] = None
    green_light: Optional[bool] = None
    notes: Optional[str] = None
    posted_at: Optional[datetime] = None
    measured_at: Optional[datetime] = None


class TrialReelOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    ghost_post_id: Optional[uuid.UUID] = None
    niche: Optional[str] = None
    variation_label: Optional[str] = None
    post_url: Optional[str] = None
    views_at_1k: Optional[int] = None
    completion_rate: Optional[float] = None
    likes: Optional[int] = None
    comments: Optional[int] = None
    shares: Optional[int] = None
    status: str
    green_light: bool
    notes: Optional[str] = None
    posted_at: Optional[datetime] = None
    measured_at: Optional[datetime] = None
    created_by: Optional[uuid.UUID] = None
    created_at: datetime


# ---------------------------------------------------------------------------
# Dashboard stats
# ---------------------------------------------------------------------------


class GhostViralityStats(BaseModel):
    total_ghost_posts: int
    new_last_7_days: int
    pattern_cards_ready: int
    active_scout_jobs: int
    trial_reels_total: int
    trial_reels_green_lit: int
    top_niches: list[dict]

File: frontend/src/api/ghostVirality.ts
import { api } from './client'

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface GhostScoutJob {
  id: string
  niche: string
  competitor_accounts: string[] | null
  status: 'pending' | 'running' | 'completed' | 'failed'
  celery_task_id: string | null
  reels_scraped: number
  ghost_viral_found: number
  error_message: string | null
  created_by: string | null
  started_at: string | null
  completed_at: string | null
  created_at: string
}

export interface PatternCard {
  id: string
  ghost_post_id: string
  hook_duration_seconds: number | null
  hook_clip_path: string | null
  scene_cut_count: number | null
  transcript_snippet: string | null
  transcript_language: string | null
  transcript_confidence: number | null
  visual_text: string | null
  audio_type: 'trending' | 'original' | 'unknown' | null
  audio_id: string | null
  audio_name: string | null
  hook_archetype: string | null
  raw_card: Record<string, unknown> | null
  created_at: string
}

export interface GhostViralPost {
  id: string
  reel_id: string
  account_username: string
  niche: string | null
  view_count: number | null
  like_count: number | null
  comment_count: number | null
  follower_count: number | null
  outlier_reach_ratio: number | null
  ghost_virality_delta: number | null
  like_percentile: number | null
  view_velocity: number | null
  like_velocity: number | null
  strategy_label: 'high_dm_share' | 'watch_time_arbitrage' | 'audio_driven' | 'polarizing' | 'utility' | 'unknown' | null
  comment_sentiment: string | null
  permalink: string | null
  thumbnail_url: string | null
  posted_at: string | null
  pattern_card_ready: boolean
  detected_at: string
  pattern_card: PatternCard | null
}

export interface NicheOverview {
  niche: string
  sample_size: number
  p10: number | null
  p30: number | null
  p50: number | null
  p70: number | null
  p90: number | null
  outlier_reach_threshold: number
  ghost_viral_count: number | null
  updated_at: string
}

export interface TrialReel {
  id: string
  ghost_post_id: string | null
  niche: string | null
  variation_label: string | null
  post_url: string | null
  views_at_1k: number | null
  completion_rate: number | null
  likes: number | null
  comments: number | null
  shares: number | null
  status: 'pending' | 'live' | 'promoted' | 'rejected'
  green_light: boolean
  notes: string | null
  posted_at: string | null
  measured_at: string | null
  created_by: string | null
  created_at: string
}

export interface GhostViralityStats {
  total_ghost_posts: number
  new_last_7_days: number
  pattern_cards_ready: number
  active_scout_jobs: number
  trial_reels_total: number
  trial_reels_green_lit: number
  top_niches: { niche: string; count: number }[]
}

// ---------------------------------------------------------------------------
// API
// ---------------------------------------------------------------------------

export const ghostViralityApi = {
  // Stats
  getStats: () => api.get<GhostViralityStats>('/ghost-virality/stats'),

  // Scout jobs
  createScoutJob: (niche: string, competitorAccounts: string[]) =>
    api.post<GhostScoutJob>('/ghost-virality/scout-jobs', {
      niche,
      competitor_accounts: competitorAccounts,
    }),
  listScoutJobs: (limit = 20) =>
    api.get<GhostScoutJob[]>(`/ghost-virality/scout-jobs?limit=${limit}`),
  getScoutJob: (id: string) =>
    api.get<GhostScoutJob>(`/ghost-virality/scout-jobs/${id}`),

  // Ghost Feed
  listGhosts: (params: {
    niche?: string
    strategy?: string
    days?: number
    limit?: number
    offset?: number
  } = {}) => {
    const q = new URLSearchParams()
    if (params.niche) q.set('niche', params.niche)
    if (params.strategy) q.set('strategy', params.strategy)
    if (params.days != null) q.set('days', String(params.days))
    if (params.limit != null) q.set('limit', String(params.limit))
    if (params.offset != null) q.set('offset', String(params.offset))
    return api.get<GhostViralPost[]>(`/ghost-virality/ghosts?${q}`)
  },
  getGhost: (id: string) => api.get<GhostViralPost>(`/ghost-virality/ghosts/${id}`),
  getPatternCard: (postId: string) =>
    api.get<PatternCard>(`/ghost-virality/ghosts/${postId}/pattern-card`),
  retryPatternRecognition: (postId: string) =>
    api.post<{ status: string; post_id: string }>(
      `/ghost-virality/ghosts/${postId}/pattern-card/retry`,
    ),

  // Niches
  listNiches: () => api.get<NicheOverview[]>('/ghost-virality/niches'),

  // Trial Reels
  createTrialReel: (data: {
    ghost_post_id?: string
    niche?: string
    variation_label?: string
    post_url?: string
    notes?: string
  }) => api.post<TrialReel>('/ghost-virality/trial-reels', data),
  listTrialReels: (niche?: string, limit = 50) => {
    const q = new URLSearchParams({ limit: String(limit) })
    if (niche) q.set('niche', niche)
    return api.get<TrialReel[]>(`/ghost-virality/trial-reels?${q}`)
  },
  getTrialReel: (id: string) => api.get<TrialReel>(`/ghost-virality/trial-reels/${id}`),
  updateTrialReel: (
    id: string,
    data: Partial<{
      post_url: string
      views_at_1k: number
      completion_rate: number
      likes: number
      comments: number
      shares: number
      status: string
      green_light: boolean
      notes: string
      posted_at: string
      measured_at: string
    }>,
  ) => api.patch<TrialReel>(`/ghost-virality/trial-reels/${id}`, data),
}
